File: ObjectCompiling/Compile.lua
local function valueInstanceFromType(value)
    if type(value) == "string" then
        return Instance.new("StringValue")
    elseif type(value) == "boolean" then
        return Instance.new("BoolValue")
    elseif type(value) == "number" then
        return Instance.new("NumberValue")
    elseif typeof(value) == "Color3" then
        return Instance.new("Color3Value")
    elseif typeof(value) == "Instance" then
        return Instance.new("ObjectValue")
    elseif typeof(value) == "CFrame" then
        return Instance.new("CFrameValue")
    elseif typeof(value) == "Vector3" then
        return Instance.new("Vector3Value")
    else
        return nil
    end
end

local function convertToConfiguration(obj, parent)
    local config = Instance.new("Configuration")
    config.Name = parent and (parent.Name .. "_" .. obj.Name) or obj.Name
    config.Parent = parent or workspace

    for key, value in pairs(obj) do
        local valueInstance = valueInstanceFromType(value)
        if valueInstance then
            valueInstance.Name = key
            valueInstance.Value = value
            valueInstance.Parent = config
        elseif type(value) == "table" then
            convertToConfiguration(value, config)
        end
    end

    return config
end

-- Example usage:
local exampleObject = {
    Name = "ExampleObject",
    StringValue = "Hello, World!",
    NumberValue = 42,
    BoolValue = true,
    NestedObject = {
        Name = "NestedObject",
        NestedNumber = 7
    }
}

local config = convertToConfiguration(exampleObject)

File: ObjectCompiling/Compile.ts
import Object from "@rbxts/object-utils";

type DataType = string | boolean | number | Color3 | Instance | CFrame | Vector3;

interface ObjectData {
	[key: string]: DataType | ObjectData;
}

function valueInstanceFromType(value: DataType, key: string): Instance | undefined {
	if (typeOf(value) === "string") {
		const stringValue = new Instance("StringValue");
		stringValue.Name = key;
		return stringValue;
	} else if (typeOf(value) === "boolean") {
		const boolValue = new Instance("BoolValue");
		boolValue.Name = key;
		return boolValue;
	} else if (typeOf(value) === "number") {
		const numberValue = new Instance("NumberValue");
		numberValue.Name = key;
		return numberValue;
	} else if (value instanceof Color3) {
		const color3Value = new Instance("Color3Value");
		color3Value.Name = key;
		return color3Value;
	} else if (value instanceof Instance) {
		const objectValue = new Instance("ObjectValue");
		objectValue.Name = key;
		return objectValue;
	} else if (value instanceof CFrame) {
		const cframeValue = new Instance("CFrameValue");
		cframeValue.Name = key;
		return cframeValue;
	} else if (value instanceof Vector3) {
		const vector3Value = new Instance("Vector3Value");
		vector3Value.Name = key;
		return vector3Value;
	} else {
		return undefined;
	}
}

export function convertToConfiguration(
	obj: ObjectData,
	Info: {
		parent: Instance;
		configName: string;
	},
): Instance {
	const config = new Instance("Configuration");
	config.Name = Info.configName;
	config.Parent = Info.parent;

	for (const key of Object.keys(obj)) {
		if (obj[key]) {
			const value = obj[key];
			const valueInstance = valueInstanceFromType(value as DataType, tostring(key));
			if (valueInstance) {
				(valueInstance as any).Value = value;
				valueInstance.Parent = config;
			} else if (typeOf(value) === "table") {
				convertToConfiguration(value as ObjectData, {
					configName: typeIs(key, "string") ? key : "ChildConfigration",
					parent: config,
				});
			}
		}
	}

	return config;
}
